csrc/scheduler/matmul.cpp

```diff
@@ -51,6 +57,377 @@ void moveInnerBroadcastLeft(TensorView* tv, int number_of_inner_pos = 3) {
+     * where init is the initial position of the pointer in the clock when we
+     * start the game, and stride is the number of ticks we move forward each
+     * time, and i is the number of times we move forward.
+     *
+     * In our problem, f(i) is the megabank of the `i`th row of the matrix, and
+     * `init` is the megabank of the 0th row of the matrix.
+     *
+     * One very important property of f(i) is:
+     * - if f(i1) == f(i2), then for every j, f(i1 + j) = f(i2 + j)
+     * This property is true because:
+     *   f(i1 + j) = f(i1) + j * stride = f(i2) + j * stride = f(i2 + j)
+     *
+     * The above property tells us, as we turn the clock forward:
+     * - initially, we will go to a never-visited tick in each turn, but,
+     * - at some point, we will return back to our original position, and,
+     * - after we return, we start repeat the pervious pattern again and again.
+     *
+     * As an example, consider f(i) where init = 0, stride = 6, under Z/8Z:
+     *     i  0 1 2 3 4 5 6 7
+     *   f(i) 0 6 4 2 0 6 4 2
+     * We can see that f(i) is repeating a pattern of four unique numbers
+     * "0 6 4 2" twice. In our bank conflict problem, this means we are using 4
+     * different megabanks, and we have a 2-way conflict.
+     *
+     * The question of interest is, does the above observation generalize? That
+     * is, does f(i) always repeat a pattern of p unique numbers q times? Note
+     * that p and q must satisfy p * q = n.
+     *
+     * The answer to the above question is: yes! Consider the following
+     * equation:
+     *    f(i1 + j) == f(i1)
+     * We want to know what is the smallest positive number j that makes the
+     * above equation true. Because this tells us in how many steps we will see
+     * repeat. This equation can be simplified as:
+     *   f(i1 + j) == f(i1) + j * stride == f(i1)
+     *   ==> j * stride == 0
+     *
+     * An important tool to study this equation is multiplicative inverse:
+     * https://en.wikipedia.org/wiki/Modular_multiplicative_inverse
+     * stride has an multiplicative inverse if and only if stride coprime with
+     * n, that is, g := gcd(stride, n) == 1. In such case, the solution to our
+     * equation j * stride == 0 is j = stride^(-1) * 0 = 0, that is: f(i) does
+     * not repeat, that is: there is no bank conflict.
+     */
+
+    int g = std::gcd(num_megabanks, row_stride_znz);
+    if (g == 1) {
+      return; // No need to swizzle in this case.
+    }
+
+    /* For the case where stride does not coprime with n, if n is a power of a
+     * prime, then we can write stride as:
+     *   stride = s * g
+     * where s coprime with n. Then the equation j * stride == 0 becomes:
+     *   j * s * g == 0
+     * which can be simplified as
+     *   j * g == s^(-1) * 0
+     *   ==> j * g == 0
+     *
+     * It is easy to see that j is n / g.
+     * That is: f(i) always repeat a pattern of n / g unique numbers g times
```

Comment:
Do you mean `g times` here is g-way bank conflicts?

Reply:
Yes, exactly. Let me add this to the comment, so that we don't get lost on which math concept correspond to what in the bank-conflicting problem.

Reply:
added